File: 0032.LongestValidParentheses.go
func longestValidParentheses(s string) int {
    maxLen := 0
    stack := []int{-1}
    for i, ch := range s {
        if ch == '(' {
            stack = append(stack, i)
        } else {
            stack = stack[:len(stack)-1]
            if len(stack) > 0 {
                maxLen = max(maxLen, i-stack[len(stack)-1])
            } else {
                stack = append(stack, i)
            }
        }
    }
    return maxLen
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: 0000-0999/0032.LongestValidParentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s:
            return 0
        #dp = [0] * (len(s)+1)
        dp = [0] * len(s)
        stack = []
        for i in xrange(len(s)):
            if s[i] == '(':
                stack.append(i)
            elif stack:
                l = stack.pop()
                #dp[i+1] = dp[l] + i - l + 1
                dp[i] = dp[l-1] + i - l + 1
        return max(dp)
